src/core/Block.ts

```diff
@@ -0,0 +1,260 @@
+import EventBus from './EventBus';
+import { nanoid } from 'nanoid';
+import Handlebars from 'handlebars';
+
+interface BlockMeta<P = any> {
+  props: P;
+}
+
+type Events = Values<typeof Block.EVENTS>;
+
+export default class Block<P = any> {
+  static EVENTS = {
+    INIT: 'init',
+    FLOW_CDM: 'flow:component-did-mount',
+    FLOW_CDU: 'flow:component-did-update',
+    FLOW_RENDER: 'flow:render',
+  } as const;
+
+  public id = nanoid(6);
+  private readonly _meta: BlockMeta;
+
+  protected _element: Nullable<HTMLElement> = null;
+  protected readonly props: P;
+  protected children: { [id: string]: Block } = {};
+
+  eventBus: () => EventBus<Events>;
+
+  protected state: any = {};
+  protected refs: { [key: string]: HTMLElement } = {};
+
+  public constructor(props?: P) {
+    const eventBus = new EventBus<Events>();
+
+    this._meta = {
+      props,
+    };
+
+    this.getStateFromProps(props);
+
+    this.props = this._makePropsProxy(props || ({} as P));
+    this.state = this._makePropsProxy(this.state);
+
+    this.eventBus = () => eventBus;
+
+    this._registerEvents(eventBus);
+
+    eventBus.emit(Block.EVENTS.INIT, this.props);
+  }
+
+  _registerEvents(eventBus: EventBus<Events>) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  _createResources() {
+    this._element = this._createDocumentElement('div');
+  }
+
+  protected getStateFromProps(props: any): void {
+    this.state = {};
+  }
+
+  init() {
+    this._createResources();
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER, this.props);
+  }
+
+  _componentDidMount(props: P) {
+    this.componentDidMount(props);
+  }
+
+  componentDidMount(props: P) {}
+
+  _componentDidUpdate(oldProps: P, newProps: P) {
+    const response = this.componentDidUpdate(oldProps, newProps);
+    if (!response) {
+      return;
+    }
+    this._render();
+  }
+
+  componentDidUpdate(oldProps: P, newProps: P) {
+    return true;
+  }
+
+  setProps = (nextProps: P) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+  setChildProps = (childRefName: string, nextProps: P) => {
+
+    if (!nextProps || !childRefName) {
+      return;
+    }
+    const childComponent = this.retrieveChildByRef(childRefName);
+    childComponent.setProps(nextProps);
+    this.refs[childRefName] = childComponent.getContent();
+  };
+
+  retrieveChildByRef = (ref: string) => {
+
+    const childBlocks = Object.values(this.children).filter(
+      (c) => c.element === this.refs[ref]
```

Comment:
Лучше не делать такие сокращения. Страдает читаемость кода.
```suggestion
    const childBlocks = Object.values(this.children).filter(
      (child) => child.element === this.refs[ref]
```